clippy_lints/src/methods/parsed_string_literals/primitive_types.rs

```diff
@@ -0,0 +1,56 @@
+use std::fmt::Display;
+use std::str::FromStr;
+
+use clippy_utils::get_parent_expr;
+use clippy_utils::source::str_literal_to_char_literal;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self, Ty};
+use rustc_span::Symbol;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, lit: Symbol, ty: Ty<'_>, strlit: &Expr<'_>) {
+    macro_rules! number {
+        ($kind:ident, $expr:expr, $msg:expr, [$($subkind:ident => $ty:ident),*$(,)?]$(,)?) => {{
+            match $expr {
+                $(ty::$kind::$subkind => (try_parse::<$ty>(lit, Some(stringify!($ty))), $msg),)*
+                #[allow(unreachable_patterns)]
+                _ => return,
+            }
+        }};
+    }
+
+    let mut app = Applicability::MachineApplicable;
+    if let (Some(subst), entity) = match ty.kind() {
+        ty::Int(int_ty) => number!(IntTy, int_ty, "a signed integer",
+                                     [Isize => isize, I8 => i8, I16 => i16, I32 => i32, I64 => i64, I128 => i128]),
+        ty::Uint(uint_ty) => number!(UintTy, uint_ty, "an unsigned integer",
+                                        [Usize => usize, U8 => u8, U16 => u16, U32 => u32, U64 => u64, U128 => u128]),
+        // FIXME: ignore `f16` and `f128` for now as they cannot use the default formatter
+        ty::Float(float_ty) => number!(FloatTy, float_ty, "a real number",
+                                           [F32 => f32, F64 => f64]),
+        ty::Bool => (try_parse::<bool>(lit, None), "a boolean"),
+        ty::Char => (str_literal_to_char_literal(cx, strlit, &mut app, false), "a character"),
+        _ => return,
+    } {
+        let subst = match subst.as_bytes()[0] {
+            // Strip trailing `+` if present, which is supported by `parse` but not in literals
+            b'+' => subst[1..].into(),
+            // Higher precedence for a method call requires that negative literals are parenthesized
+            b'-' if let Some(ExprKind::MethodCall(_, recv, _, _)) = get_parent_expr(cx, expr).map(|e| e.kind)
+                && expr.hir_id == recv.hir_id =>
+            {
+                format!("({subst})")
+            },
+            _ => subst,
+        };
+        super::maybe_emit_lint(cx, expr, false, entity, subst, app);
+    }
+}
+
+fn try_parse<T: FromStr + Display>(lit: Symbol, suffix: Option<&str>) -> Option<String> {
+    lit.as_str()
+        .parse::<T>()
+        .ok()
+        .map(|_| suffix.map_or_else(|| lit.to_string(), |suffix| format!("{lit}_{suffix}")))
```

Comment:
Always adding the suffix isn't really great. Should do the same thing as ip addresses and only add it when necessary.